sim/core/resource_bar.go

```diff
@@ -0,0 +1,143 @@
+	Reset(sim *Simulation)                                             // Resets the current resource bar
+	Value() float64                                                    // Returns the current amount of resource
+}
+
+type SecondaryResourceConfig struct {
+	Type    SecondaryResourceType // The type of resource the bar tracks
+	Max     float64               // The maximum amount the bar tracks
+	Default float64               // The default value this bar should be initialized with
+}
+
+type _SecondaryResourceBar struct {
+	config  SecondaryResourceConfig
+	value   float64
+	unit    *Unit
+	metrics map[ActionID]*ResourceMetrics
+}
+
+// CanSpend implements SecondaryResourceBar.
+func (bar *_SecondaryResourceBar) CanSpend(limit float64) bool {
+	return bar.value >= limit
+}
+
+// Gain implements SecondaryResourceBar.
+func (bar *_SecondaryResourceBar) Gain(amount float64, action ActionID, sim *Simulation) {
+	bar.value += amount
+	baseAmount := amount
+	if bar.value > bar.config.Max {
+		amount -= bar.value - bar.config.Max
+		bar.value = bar.config.Max
+	}
+
+	metrics := bar.GetMetric(action)
+	metrics.AddEvent(baseAmount, amount)
+	if sim.Log != nil {
+		bar.unit.Log(
+			sim,
+			"Gained %0.0f generic resource from %s (%0.0f --> %0.0f) of %0.0f total.",
+			amount,
+			action,
+			bar.value-amount,
+			bar.value,
+			bar.config.Max,
+		)
+	}
+}
+
+// Reset implements SecondaryResourceBar.
+func (bar *_SecondaryResourceBar) Reset(sim *Simulation) {
+	bar.value = 0
+	if bar.config.Default > 0 {
+		bar.Gain(bar.config.Default, ActionID{SpellID: int32(bar.config.Type)}, sim)
+	}
+}
+
+// Spend implements SecondaryResourceBar.
+func (bar *_SecondaryResourceBar) Spend(amount float64, action ActionID, sim *Simulation) {
+	if amount > bar.value {
+		panic("Trying to spend more resource than is available.")
+	}
+
+	metrics := bar.GetMetric(action)
+	if sim.Log != nil {
+		bar.unit.Log(
+			sim,
+			"Spent %0.0f generic resource from %s (%0.0f --> %0.0f) of %0.0f total.",
+			amount,
+			metrics.ActionID,
+			bar.value,
+			bar.value-amount,
+			bar.config.Max,
+		)
+	}
+
+	metrics.AddEvent(-amount, -amount)
+	bar.value -= amount
+}
+
+// SpendUpTo implements SecondaryResourceBar.
+func (bar *_SecondaryResourceBar) SpendUpTo(limit float64, action ActionID, sim *Simulation) float64 {
+	if bar.value > limit {
+		bar.Spend(limit, action, sim)
+		return limit
+	}
+
+	max := math.Floor(bar.value)
+	bar.Spend(max, action, sim)
+	return max
+}
+
+// Value implements SecondaryResourceBar.
+func (bar *_SecondaryResourceBar) Value() float64 {
+	return bar.value
+}
+
+func (bar *_SecondaryResourceBar) GetMetric(action ActionID) *ResourceMetrics {
+	metric, ok := bar.metrics[action]
+	if !ok {
+		metric = bar.unit.NewGenericMetric(action)
+		bar.metrics[action] = metric
+	}
+
+	return metric
+}
+
+func (unit *Unit) NewSecondaryResourceBar(config SecondaryResourceConfig) SecondaryResourceBar {
```

Comment:
I would recommend merging this with `RegisterSecondaryResourceBar()` so that there is only one public access method containing all the validation checks.